From 4dfd00a83d5e4cd0c588aa378d910cbd1fdf24f0 Mon Sep 17 00:00:00 2001
From: Matthias Brugger <mbrugger@suse.com>
Date: Tue, 25 Apr 2017 17:57:50 +0200
Subject: [PATCH] kexec: arm64: Add kernel parameter to enable kexec

Patch-mainline: Never, this is our internal watchguard.
References: fate#320671

On arm64 we need to enable kexec explicitly via a kernel
boot parameter.

Signed-off-by: Matthias Brugger <mbrugger@suse.com>
---
 kernel/kexec.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/kernel/kexec.c b/kernel/kexec.c
index 28063711a101..4878d8ec825d 100644
--- a/kernel/kexec.c
+++ b/kernel/kexec.c
@@ -21,6 +21,17 @@
 
 #include "kexec_internal.h"
 
+#ifdef CONFIG_ARM64
+bool kexec_enabled = false;
+static int __init set_kexec_enabled(char *str)
+{
+	return strtobool(str, &kexec_enabled);
+}
+__setup("kexec=", set_kexec_enabled);
+#else
+bool kexec_enabled = true;
+#endif
+
 static int copy_user_segment_list(struct kimage *image,
 				  unsigned long nr_segments,
 				  struct kexec_segment __user *segments)
@@ -132,7 +143,7 @@ SYSCALL_DEFINE4(kexec_load, unsigned long, entry, unsigned long, nr_segments,
 	int result;
 
 	/* We only trust the superuser with rebooting the system. */
-	if (!capable(CAP_SYS_BOOT) || kexec_load_disabled)
+	if (!capable(CAP_SYS_BOOT) || kexec_load_disabled || !(kexec_enabled))
 		return -EPERM;
 
 	if (get_securelevel() > 0)
-- 
2.12.0

